Experimental wasm support

bluenote10 opened the pull request:
Regarding #65, I think I've found a way to actually execute at least one of the existing test suites with WASM without too much of a change, and things look actually quite good on first glance:

![image](https://github.com/ExpHP/npyz/assets/3620703/89a57ce1-66c4-4d04-ae39-5313caf868fb)

I've tried adding it to the CI pipeline, but it may require a bit of tweaking until it works there I guess...

ExpHP commented:
LGTM!

Yes, we definitely want CI to be testing this.  I made a change that I think should fix the CI not running on pull requests, but it still hasn't run.  Maybe a rebase on the updated `master` & force push will make it fire?

ExpHP commented:
Actually, it would probably be better to run a greater subset of tests.  I am wondering, which tests do not work in WASM?  If it's only a few maybe we could blacklist them somehow.  (I am thinking possibly via a short shell script)

bluenote10 commented:
> Actually, it would probably be better to run a greater subset of tests. I am wondering, which tests do not work in WASM?

The issue is that running tests against WASM is quite different than running regular tests. The biggest differences are:
- File I/O isn't possible, which is used in many existing tests. I've changed one of them to use `include_bytes!` instead as an example of how one could run it, but it is slightly awkward of course to pull it into the binary.
- One other difference is that the tests must be in the root of the crate, or within a `pub mod`. Putting them inside a private module will not work. That is quite unfortunate in my opinion, because often the high-quality unit tests are implemented locally inside a private module. I'm not sure if having public test modules is a good idea (I would be worried that the are considered part of the public interface, but I'm not sure if the `#[cfg(test)]` would actually prevent that).
- Anything that tries to link against native libraries (here zlib) doesn't work. So, all tests that address compression wouldn't work.
- `println!` doesn't work, which can be annoying if tests rely on producing helpful debug information, but this doesn't seem to be the case here.

In general I'd consider the trick `use wasm_bindgen_test::wasm_bindgen_test as test` a bit experimental, because it requires some care to write tests that work under both. At least the [wasm-bindgen-test](https://crates.io/crates/wasm-bindgen-test) crate documents the `#[wasm_bindgen_test]` macro as:

> This is the runtime support needed to execute tests. This is basically the same thing as the test crate in the Rust repository, and _one day it will likely use the test crate itself_! For now though it's a minimal reimplementation that provides the support for [...]


I've enabled a few more tests now as an example. Do you have any other tests in mind that would be particularly valuable?


> I am thinking possibly via a short shell script

I've added a small script. Any preference where it should live?

bluenote10 commented on tests/serialize_array.rs:
```diff
@@ -132,9 +135,12 @@ fn default_dtype() {
     use super::*;
 
+    #[cfg(target_arch="wasm32")]
+    use wasm_bindgen_test::wasm_bindgen_test as test;
```
For some reason it was necessary to repeat that alias here.

bluenote10 commented:
> I made a change that I think should fix the CI not running on pull requests, but it still hasn't run.

I think you'll have to approve the workflow execution once, because it is the first time that I'm making a contribution to the repo. Then it should run now.

ExpHP commented:
AHAH, `curl` was getting piped into `sh ./run_wasm_tests.sh`, because that's how multiline strings with no delimiter in yaml get parsed.

ExpHP commented:
The main thing I would like to test is really the unittests in the library, which can normally be done using `cargo test --lib`, but [that doesn't seem to work with wasm-pack](https://github.com/ExpHP/npyz/issues/65#issuecomment-1577434904) so I'm not sure...

I feel bad for the delay, so I went ahead and released 0.8.1.
